### PR TITLE
Allow specifying the `local-disk` as a sized `SSD`/`HDD`.

### PR DESCRIPTION
**Pull request by kshakir**

### CODE REVIEW

**Comment by kshakir** [comment]
First reviewer @scottfrazer

File: src/main/java/cromwell/binding/DiskType.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package cromwell.binding;
+
+public enum DiskType {
```
Comment (geoffjentry) [inline src/main/java/cromwell/binding/DiskType.java]: What is this giving you that an ADT wouldn't?

File: src/main/scala/cromwell/binding/RuntimeAttributes.scala [hunk 2]
```diff
@@ -204,7 +219,7 @@ object RuntimeAttributes {
     try {
       value.toLong.successNel
     } catch {
-      case _: IllegalArgumentException => s"$value not convertable to a Long".failureNel[Long]
+      case _: IllegalArgumentException => s"$value not convertible to a Long".failureNel[Long]
```
Comment (geoffjentry) [inline src/main/scala/cromwell/binding/RuntimeAttributes.scala]: thank you for pointing out my horrendous spelling ;)

File: src/main/java/cromwell/binding/DiskType.java (hunk 1 above)
Comment (kshakir) [inline src/main/java/cromwell/binding/DiskType.java]: Built in `DiskType.values`. And never been a fan of the limited functionality in `scala.Enumeration`.

File: src/main/scala/cromwell/binding/RuntimeAttributes.scala (hunk 2 above)
Comment (kshakir) [inline src/main/scala/cromwell/binding/RuntimeAttributes.scala]: Thank `F2` in IntelliJ. I always look for its clean green :white_check_mark:.

File: src/main/java/cromwell/binding/DiskType.java (hunk 1 above)
Comment (geoffjentry) [inline src/main/java/cromwell/binding/DiskType.java]: agree on the uselessness of enumeration. only asked as when i made a quick skim I didn't see any Enum-ness being used but emphasis on quick skim :)

File: src/main/scala/cromwell/binding/RuntimeAttributes.scala [hunk 3]
```diff
@@ -107,19 +109,32 @@ object RuntimeAttributes {
     value match {
       case Some(v) =>
         val nels = v.split(",\\s*").toList map { x => validateLocalDisk(x) }
```
Comment (mcovarr) [inline src/main/scala/cromwell/binding/RuntimeAttributes.scala]: not part of your changes, but
```scala
expr.toList map validateLocalDisk
```

File: src/main/scala/cromwell/binding/RuntimeAttributes.scala [hunk 4]
```diff
@@ -107,19 +109,32 @@ object RuntimeAttributes {
...
-        // Use the LocalizationDisk as the start of the fold. Currently same as default, but perhaps not always
-        val defaultDiskNel = Vector(LocalizationDisk).successNel[String]
-        nels.foldLeft(defaultDiskNel)((acc, v) => (acc |@| v) { (a, v) => a :+ v })
+        val emptyDiskNel = Vector.empty[Disk].successNel[String]
+        val disksNel = nels.foldLeft(emptyDiskNel)((acc, v) => (acc |@| v) { (a, v) => a :+ v })
+        disksNel map {
+          case disks if disks.exists(disk => disk.getName == LocalDiskName) => disks
```
Comment (mcovarr) [inline src/main/scala/cromwell/binding/RuntimeAttributes.scala]: `_.getName == LocalDiskName`

File: src/main/scala/cromwell/binding/RuntimeAttributes.scala [hunk 5]
```diff
@@ -230,8 +245,12 @@ object RuntimeAttributes {
     astList.asScala.toVector map { a => processRuntimeAttribute(a.asInstanceOf[Ast]) } toMap
   }
 
-  private case class LocalDisk(name: String, sizeGb: Long, diskType: String) {
-    def toDisk: Disk = new Disk().setSizeGb(sizeGb).setType(diskType).setName(name)
+  private case class LocalDisk(name: String, diskType: DiskType, sizeGbOption: Option[Long] = None) {
+    def toDisk: Disk = {
+      val disk = new Disk().setName(name).setType(diskType.googleTypeName).setAutoDelete(true)
+      sizeGbOption.foreach(sizeGb => disk.setSizeGb(sizeGb)) // scala.Long to java.lang.Long
```
Comment (mcovarr) [inline src/main/scala/cromwell/binding/RuntimeAttributes.scala]:
```scala
sizeGbOption foreach disk.setSizeGb
```

**Comment by mcovarr** [comment]
:+1:

**Comment by scottfrazer** [comment]
I was definitely stoked to see this PR this morning, considering I need this!  Anyway there are a few things I need some clarification on:

1.  Is `local-disk` meant to refer to the disk that we will be used as the current working directory of the process?
2.  If you mount a second disk, how do you know where it's mounted on the host?

File: src/main/scala/cromwell/binding/RuntimeAttributes.scala (hunk 5 above)
Comment (kshakir) [inline src/main/scala/cromwell/binding/RuntimeAttributes.scala]: This one doesn't work unless I do more conversion elsewhere. sizeGb is a `scala.Long`. `setSizeGb` takes a `java.lang.Long`.

Comment (kshakir) [inline src/main/scala/cromwell/binding/RuntimeAttributes.scala]: Intellij automagic to the rescue again.

`sizeGbOption map long2Long foreach disk.setSizeGb`

**Comment by kshakir** [comment]
@scottfrazer 

Let's discuss off github if you'd like to go in depth. There's more background in the original JES docs too.

In short for now:

1. I can't get all the "logic" terminology correct here (a Fruit is not necessarily an Apple), but shared variously between `RuntimeAttributes` and `JesBackend`, `local-disk` is `LocalWorkingDiskValue` is `working_disk` is `/cromwell_root`.
2. Additional mount points are not input or captured anywhere within cromwell. So, if you _try to_ mount a second disk, cromwell never finishes setting the JES parameters for where to mount the freshly attached drive.

This PR tries not to break this existing "functionality", while also allowing one to effectively increase the size of `/cromwell_root`. I'm leaving it to another ticket to discuss how this should be enhanced.

**Comment by scottfrazer** [comment]
Yes I think we do need to discuss... I think the issue I'm having in JES is that the jobs are writing all their outputs to `/` which only has 10GB

**Comment by kshakir** [comment]
The dirtiest solution I've seen in the past is a good 'ol `cd /cromwell_root && \` at the top of whatever exec script.

But let's chat.

**Comment by kshakir** [comment]
@scottfrazer The latest commit should localize to, cd to, and register relative outputs from `/cromwell_root`.

**Comment by scottfrazer** [comment]
cool, I'm looking now

File: src/main/scala/cromwell/engine/backend/jes/JesBackendCall.scala [hunk 6]
```diff
@@ -1,11 +1,10 @@
 package cromwell.engine.backend.jes
 
-import java.io.File
 import java.nio.file.Paths
 
 import cromwell.binding._
 import cromwell.binding.values.WdlFile
-import cromwell.engine.backend.jes.JesBackend.{JesInput, JesOutput}
+import cromwell.engine.backend.jes.JesBackend.{JesOutput, localFilePathFromRelativePath}
```
Comment (scottfrazer) [inline src/main/scala/cromwell/engine/backend/jes/JesBackendCall.scala]: I usually don't see imports of functions in our codebase... is that considered good practice in Scala?

**Comment by scottfrazer** [comment]
:+1: Should there be another ticket filed to capture the ability to specify where to mount these disks?

File: src/main/scala/cromwell/engine/backend/jes/JesBackendCall.scala (hunk 6 above)
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/backend/jes/JesBackendCall.scala]: Opinions i've seen vary as wildly as they do in java-land, which is to say there are as many opinions on imports as there are users

**Pull request merged**